artemis-commons/src/main/java/org/apache/activemq/artemis/api/core/SimpleString.java

```diff
@@ -259,22 +281,23 @@ public boolean equals(final Object other) {
       if (other instanceof SimpleString) {
          SimpleString s = (SimpleString) other;
 
-         if (data.length != s.data.length) {
-            return false;
-         }
-
-         for (int i = 0; i < data.length; i++) {
-            if (data[i] != s.data[i]) {
-               return false;
-            }
-         }
-
-         return true;
+         return ByteUtil.equals(data, s.data);
```

Comment:
I've implemented it in the last commit as:
```
   @Override
   public boolean equals(final Object other) {
      if (this == other) {
         return true;
      }

      if (other instanceof SimpleString) {
         SimpleString s = (SimpleString) other;
         if (s.hash != 0 && this.hash != 0) {
            if (s.hash != this.hash)
               return false;
         }
         if (s.str != null && this.str != null) {
            return this.str.equals(s.str);
         } else {
            return ByteUtil.equals(this.data, s.data);
         }
      } else {
         return false;
      }
   }
```

Reply:
I don't see the benefit of this change, for GC.

Why i moved the equals code between two byte arrays into ByteUtil and re-used the equals, was to keep the logic that is checked equal between the bytebuf equals and the equals of the String the same called code. Adding these extra checks, could allow for differences in result now, and for very marginal improvement at best IMO.

Reply:
The benefit is not from the GC pov but for normal `SimpleString` usage inside Map/Set: if the `SimpleString` has already hash computed or it is string-based it won't perform array comparison (optimzed ones for sure) but will benenfit of `String::equals` that is intrinsified by the JVM.

Reply:
It is just a further improvement, but I'm already happy with the code as it is now :+1:

Reply:
And there is a correctness issue around it (in my code too!!!): 
if `s` is `null` `s.data` will throw NPE!

Reply:
Cool. lets leave it then :)

Reply:
I will perform some benchs to see if it will impact positively on checks and/or pattern matching based on equals: only in that case I will send a separate improvement PR :+1: